### PR TITLE
Core: Memory Fixes

### PR DESCRIPTION
**Pull request by StevenMiller123**
This PR makes our memory code more robust, preventing a variety of exceptions and strange behavior I observed when using my recently shared memory stress test homebrew on shadPS4.

- Fixes `OrbisVirtualQueryInfo` struct. My homebrew was reading garbage data with the BitFields and array in place. I've replaced the BitField class use with C-style bitfields, and the std::array with a primitive char array instead.
- Proper low address handling in `sceKernelVirtualQuery`. Our previous code didn't check if the inputted address was below our vma map, which led to OOB memory accesses. My edge case error returns are based on hardware tests here.
- Edge case handling for `SearchFree`. Our previous code encountered unpredictable behavior if the inputted address was too high, or if there were no free memory areas to map.
- Changed default nameless memory mapping behavior. On real hardware, nameless mappings will have the name "anon" followed by the code address responsible for the memory call. For simplicity sake, I just used "anon" as the default name instead.
- Fixes a variety of broken memory asserts, caused by a fundamental misunderstanding of how `FindVMA` and `FindDmemArea` behave.
- Fixes a long unnoticed regression caused by #2080 
- Fixes an issue where `MapMemory` would assert when the area to map was only partially reserved.

With these fixes, my memory homebrew now executes all tests on shadPS4. Many of the returned results are still inaccurate, but these are due to deeper issues that I believe are out of scope for this PR.

Currently:
- Regresses EA SPORTS™ UFC® (CUSA00264) 
     - The only reason this game works in main is due to a bug in our code. Fixing this bug broke the game, and attempting to properly fix the regression broke Windows support.
- Improves Evolve (CUSA00432)
- Improves Cars 3: Driven to Win (CUSA07027)
- Improves Shadow of the Tomb Raider (CUSA10872)
- Improves The Order: 1886 (CUSA00076)
- Improves Final Fantasy XV (CUSA01633)

### CODE REVIEW

File: src/core/memory.cpp [hunk 1]
```diff
@@ -585,12 +609,14 @@ int MemoryManager::VirtualQuery(VAddr addr, int flags,
...
-    info->is_pooled.Assign(vma.type == VMAType::PoolReserved || vma.type == VMAType::Pooled);
-    info->is_committed.Assign(vma.IsMapped());
-    vma.name.copy(info->name.data(), std::min(info->name.size(), vma.name.size()));
+    info->is_flexible = vma.type == VMAType::Flexible ? 1 : 0;
+    info->is_direct = vma.type == VMAType::Direct ? 1 : 0;
+    info->is_stack = vma.type == VMAType::Stack ? 1 : 0;
+    info->is_pooled = vma.type == VMAType::PoolReserved || vma.type == VMAType::Pooled ? 1 : 0;
+    info->is_committed = vma.IsMapped() ? 1 : 0;
+
+    strcpy(info->name, vma.name.data());
```
Comment (squidbus) [inline src/core/memory.cpp]: Prefer `strncpy` with maximum length of `info->name`.

**Comment by GHU7924** [comment]
If I understand correctly, `The Order: 1886` is suitable for testing this PR, but if I'm wrong, then you immediately tell me so that I don't spam unnecessary tests.

There are lines `[Kernel.Vmm]` in the journal, so I thought the game was suitable.

`Main` - (build eb09c4c) - [Debug] <Critical> memory.cpp:331 operator(): Assertion Failed!

`PR` - (build 511063b) - [Debug] <Critical> memory.cpp:339 operator(): Assertion Failed!

[CUSA00076.log](https://github.com/user-attachments/files/20013411/CUSA00076.log)

**Comment by StevenMiller123** [comment]
@GHU7924 
This PR probably won't help for any memory asserts.
Most games running into the issues I fixed here would either crash without errors, or hang.

**Comment by Missake212** [comment]
EA SPORTS UFC crashes even earlier on this PR, nothing shows up in the log though

[CUSA00264.log](https://github.com/user-attachments/files/20013496/CUSA00264.log)

**Comment by GHU7924** [comment]
@StevenMiller123 Okay, I get it.

File: src/core/memory.cpp (hunk 1 above)
Comment (StevenMiller123) [reply src/core/memory.cpp]: Should be resolved now.

**Comment by StevenMiller123** [comment]
New push should fix the UFC regression @Missake212 
I've confirmed that increasing our address space doesn't break anything on Windows or Linux, but I'd like someone on Mac to test just in case.

**Comment by GHU7924** [comment]
@StevenMiller123 Build 47d4d66 

Personally, as soon as I try to launch any game, the emulator closes. Logs in the log folder are not created at all.

https://github.com/user-attachments/assets/14ddc867-f2cf-4e6d-83c6-c606e95fd61d

Previous build 18410ad works.

**Comment by StevenMiller123** [comment]
Yeah, I'm receiving similar reports on the Discord server too. I can't reproduce the issue on my end, so it might take some time for me to fix this.

**Comment by squidbus** [comment]
It’s almost certainly going to be the increase to the address space bound

**Comment by StevenMiller123** [comment]
Reverting that change for now then.

**Comment by GHU7924** [comment]
Now all games start. Build 3475a9e 

`The Order: 1886` - from `[Debug] memory.cpp:339 operator(): Assertion Failed!` 
to `[Debug] <Critical> memory.cpp:342 operator(): Assertion Failed!`
(Maybe it's not important, but I'm writing just in case)

I don't know what it's connected with, but this PR and the neighboring PR 2877 seem to more often cause graphic artifacts and vertex explosions in `The Last Guardian`, although if at the moment this behavior of the game is considered normal, then don't pay attention.

**Comment by Missake212** [comment]
No regressions on my end on the latest commit (beside the EA SPORTS UFC one we talked about on Discord 😁).

**Comment by StevenMiller123** [comment]
Pushed another round of fixes, mainly targeting asserts that never actually triggered due to the mind-numbing logic of iterators.
Now UFC should crash on an assert instead of causing an exception.

**Comment by Missake212** [comment]
EA SPORTS UFC now hits the following assert:

``[Debug] <Critical> memory.cpp:755 operator(): Assertion Failed!``
``Input address 0x70000000000 is out of bounds``

[CUSA00264.log](https://github.com/user-attachments/files/20098716/CUSA00264.log)

Cars 3 still crashes but not on a memory assert like main anymore, there's nothing in the log but it used to crash on:

``[Debug] <Critical> memory.cpp:488 operator(): Assertion Failed!``
``Provided address is not mapped``

[CUSA07027.log](https://github.com/user-attachments/files/20098811/CUSA07027.log)

Finally it seems like Tekken 7 is more prone to crashing on this PR, it can still enter the really long loading to get in game but I had to give it a few tries (around 5/10) in order to get it, other tries were silent crashes. Beside that all my other games work the same as on main.

**Comment by GHU7924** [comment]
The Order: 1886

Main

`[Debug] <Critical> memory.cpp:331 operator(): Assertion Failed!`

PR (build 5c5dcf0)

`[Debug] <Critical> memory.cpp:354 operator(): Assertion Failed!`
`Memory region  0x20a200000 to 0x20a400000 isn't free enough to map region 0x20a200000 to 0x20a400000`

[CUSA00076.log](https://github.com/user-attachments/files/20102168/CUSA00076.log)

All other games work without changes, that is, the same as in Main.

**Comment by StevenMiller123** [comment]
@GHU7924 Could you test The Order: 1886 with my new push?
I added some extra logging that might help debug that problem.

**Comment by GHU7924** [comment]
@StevenMiller123 I saw that the last few lines had changed, but the error remained the same.

[Kernel.Vmm] <Info> memory.cpp:157 sceKernelReserveVirtualRange: in_addr = 0x0, out_addr = 0x20a200000, len = 0x98000, flags = 0x80, alignment = 0x200000
[Kernel.Vmm] <Info> memory.cpp:70 sceKernelAllocateDirectMemory: searchStart = 0x0, searchEnd = 0x120000000, len = 0x200000, alignment = 0x200000, memoryType = 0x0, physAddrOut = 0x76c00000
[Kernel.Vmm] <Info> memory.cpp:168 sceKernelMapNamedDirectMemory: in_addr = 0x20a200000, len = 0x200000, prot = 0x3, flags = 0x10, directMemoryStart = 0x76c00000, alignment = 0x0, name = 'anon'
`[Debug] <Critical> memory.cpp:354 operator(): Assertion Failed!`
`Memory region  0x20a200000 to 0x20a400000 isn't free enough to map region 0x20a200000 to 0x20a400000`

**Comment by StevenMiller123** [comment]
Alright, it's the issue I expected then. Thanks for the test.
That particular bug is one I'll try fixing in a future PR, this one's getting pretty wide-reaching at this point.

**Comment by bax-cz** [comment]
**Minecraft 02.92 (CUSA00265)**
[Kernel.Vmm] <Info> memory.cpp:70 sceKernelAllocateDirectMemory: searchStart = 0x0, searchEnd = 0x128000000, len = 0x4000, alignment = 0x4000, memoryType = 0x3, physAddrOut = 0x129ec000
[Kernel.Vmm] <Info> memory.cpp:207 sceKernelMapDirectMemory: called, redirected to sceKernelMapNamedDirectMemory
[Kernel.Vmm] <Info> memory.cpp:168 sceKernelMapNamedDirectMemory: in_addr = 0x0, len = 0x4000, prot = 0x33, flags = 0x0, directMemoryStart = 0x129ec000, alignment = 0x4000, name = 'anon'
**[Debug] <Critical> address_space.cpp:194 operator(): Assertion Failed!
Attempt to map 0x144eb0000 with size 0x4000 which overlaps with 0x144eb0000 mapping**
[shad.log](https://github.com/user-attachments/files/20109207/shad.log)

**Comment by StevenMiller123** [comment]
@bax-cz Please share a full log. This little chunk of the log doesn't tell me anything helpful.

**Comment by MajorP93** [comment]
> @bax-cz Please share a full log. This little chunk of the log doesn't tell me anything helpful.

I think the full log got added by him, there is a link at the end of his post

**Comment by StevenMiller123** [comment]
I'll see if I can reproduce the issue on my dump of Minecraft, though the newest I can dump is v2.78.

**Comment by Missake212** [comment]
@StevenMiller123 doesn't happen on v2.78.

[CUSA00265.log](https://github.com/user-attachments/files/20109793/CUSA00265.log)

**Comment by StevenMiller123** [comment]
@Missake212 I'm unable to reproduce instabilities in any of my Unreal Engine games, could you provide a log from one of Tekken 7's crashes?

**Comment by Missake212** [comment]
> @Missake212 I'm unable to reproduce instabilities in any of my Unreal Engine games, could you provide a log from one of Tekken 7's crashes?

I'll give you a log and the call stack since the crash doesn't really show up in the log.

**Comment by Missake212** [comment]
Forget it, I can't reproduce it either, I guess I just got unlucky 8 times in a row earlier today when testing it.

**Comment by StevenMiller123** [comment]
@Missake212 Cars 3 might improve with this new push.
Turns out a change I made in https://github.com/shadps4-emu/shadPS4/pull/2080 was corrupting the VMA map.

**Comment by Missake212** [comment]
Gets in game now but unfortunately nothing really renders.

image

[CUSA07027.log](https://github.com/user-attachments/files/20114252/CUSA07027.log)

**Comment by StevenMiller123** [comment]
@GHU7924 New push should fix `The Order: 1886`, let me know how it goes when you have a chance to test.
I figured I'd handle it since my current fixes (more so the things I've learned while making these fixes) made it a easy task.

**Comment by StevenMiller123** [comment]
Oops, broke things.

**Comment by StevenMiller123** [comment]
Now it's worth testing.

**Comment by diegolix29** [comment]
Shadow of the tomb raider progress further on this pr 
[shad_log.txt](https://github.com/user-attachments/files/20114949/shad_log.txt)

**Comment by diegolix29** [comment]
[shad_log.txt](https://github.com/user-attachments/files/20114986/shad_log.txt)

here is the log without isdevkit enable.

**Comment by Graubek** [comment]
The Order: 1886 seems to hang without any errors (just black screen):  [shad_log.txt](https://github.com/user-attachments/files/20118333/shad_log.txt)
But when validation layers are turned on it shows 
```
[Render.Vulkan] <Error> vk_platform.cpp:56 DebugUtilsCallback: CDL: Invalid Command Buffer Usage: Reset of VkCommandBuffer in use by GPU: CommandPool: Command Buffer 0 (0x000000001260FE70)
[Render.Vulkan] <Error> vk_platform.cpp:56 DebugUtilsCallback: CDL: InvalidCommandBuffer
```
Val layers: [shad_log.txt](https://github.com/user-attachments/files/20118362/shad_log.txt)

**Comment by GHU7924** [comment]
@StevenMiller123 Build e41270f

When I launch `The Order: 1886`, my processor load increases sharply to 90% (namely the shadps4 application), RAM ~787 Mb and gradually increases.

This is what you can see on the screen all the time:



[CUSA00076.log](https://github.com/user-attachments/files/20120268/CUSA00076.log)

That is, the log does not go further, but the game continues to work, the FPS counter also functions, and there is no crash or closing of the emulator.

**Comment by StevenMiller123** [comment]
Thanks for the test, my code is probably hanging.
I'll see if I can reproduce and fix this when I have a chance.

**Comment by StevenMiller123** [comment]
@GHU7924 Let me know if the newest push I've made fixes this hang.

**Comment by Graubek** [comment]
The Order is now displaying splashscreen but crashes after it
`[Debug] <Critical> emit_spirv_context_get_set.cpp:169 operator(): Assertion Failed!`

[shad_log.txt](https://github.com/user-attachments/files/20124558/shad_log.txt)

obraz

**Comment by StevenMiller123** [comment]
Looks like that's another test case for that DMA PR then.
Just gotta get this merged first.

**Comment by StevenMiller123** [comment]
I think I've fixed enough issues to outweigh the UFC regression.
Marking as ready for review.

**Comment by GHU7924** [comment]
@StevenMiller123 

`[Debug] <Critical> emit_spirv_context_get_set.cpp:169 operator(): Assertion Failed!`

[CUSA00076.log](https://github.com/user-attachments/files/20124656/CUSA00076.log)

I confirm, everything is the same as with @Graubek.

File: src/core/libraries/kernel/memory.cpp [hunk 2]
```diff
@@ -126,17 +126,11 @@ s32 PS4_SYSV_ABI sceKernelAvailableDirectMemorySize(u64 searchStart, u64 searchE
...
-    if (!addr) {
-        return ORBIS_KERNEL_ERROR_EACCES;
-    }
     auto* memory = Core::Memory::Instance();
     return memory->VirtualQuery(std::bit_cast<VAddr>(addr), flags, info);
 }
 
 s32 PS4_SYSV_ABI sceKernelReserveVirtualRange(void** addr, u64 len, int flags, u64 alignment) {
-    LOG_INFO(Kernel_Vmm, "addr = {}, len = {:#x}, flags = {:#x}, alignment = {:#x}",
-             fmt::ptr(*addr), len, flags, alignment);
```
Comment (squidbus) [inline src/core/libraries/kernel/memory.cpp:138]: If this is moved to capture the out address, would prefer to keep this here and then log the out address separately, so we can see this in any error case that we may not reach the end.

Comment (StevenMiller123) [reply src/core/libraries/kernel/memory.cpp:138]: Fixed with my newest commit.
